### PR TITLE
Add typing to utils

### PR DESCRIPTION
**Pull request by dstansby**
Because apparently I type for fun now?

### CODE REVIEW

**Review by Cadair — COMMENTED** [review]
Thanks for the PR! I would not be adverse to eventually having full typing support in parfive I think it would make some things easier to reason with.

Apart from the runtime `aioftp` issue, my comments are mainly my inexperience with typing and my general frustration at what feels like weird work arounds.

File: parfive/utils.py [hunk 1]
```diff
@@ -3,11 +3,14 @@
 import hashlib
 import pathlib
 import warnings
+from typing import Any, Dict, List, Tuple, Union, TypeVar, Generator
 from pathlib import Path
 from itertools import count
 from concurrent.futures import ThreadPoolExecutor
 
+import aioftp
```
[the remark this answers is not in the document]
Comment (dstansby) [reply parfive/utils.py:11]: 👍 I've put this import under a `TYPE_CHECKING` guard so it's only imported when type checking.

File: parfive/utils.py [hunk 2]
```diff
@@ -99,53 +102,53 @@ async def get_ftp_size(client, filepath):
...
     return int(size) if size else size
 
 
-def replacement_filename(path):
+def replacement_filename(path: os.PathLike[str]) -> Path:  # type: ignore[return]
     """
     Given a path generate a unique filename.
     """
-    path = pathlib.Path(path)
+    path_ = pathlib.Path(path)
```
[the remark this answers is not in the document]
Comment (dstansby) [reply parfive/utils.py]: I think any change in type is an error 😢 . I'll just put a type ignore comment after it, I agree that this is a bit of a pain. Edit: turns out this doesn't matter until/if mypy is set to strict, so I just undid these variable renaming changes.

**Review by Cadair — APPROVED** [review]

**Comment by Cadair** [comment]
Other than the fact that the mypy ci is angry it looks good.

**Pull request merged**